proposals/0000-template-patterns.rst

```diff
@@ -0,0 +1,376 @@
+::
+
+	{-# RULES "wombat"  forall f x.  foo x (\y. f y) = bar x f  #-}
+
+* 	**Template**.
+	The LHS of a rule is called its *template*.
+* 	**Template variables**.
+	The ``forall``'d variables are called the *template variables*.
+	In rule "wombat", ``f`` and ``x`` are template variables.
+* 	**Local binders**.
+	The *local binders* of a rule are the variables bound inside the template.
+	Example: ``y`` is a local binder of rule "wombat".
+	A local binder is specifically not a template variable, nor is it free in the entire rule.
+* 	**Target**.
+	The rule matcher matches the LHS of the rule (the template) against an expression in the program (the *target*).
+* 	**Substitution**.
+	A sucessful match finds a *substitution* S: a binding for each template variable, such that applying S to the LHS yields the target.
+* 	After a successful match we replace the target expression with the substitution S applied to the RHS of the rule.
+
+
+In GHC today, a template variable ``v`` matches any expression ``e`` if
+
+* ``e`` has the same type as ``v``
+* No local binder of the template is free in ``e``.
+
+The change proposed here is that a **higher order pattern** matches any expression (of the same type):
+
+* 	**Higher order pattern (HOP)**.
+	A *higher order pattern* is an expression of form ``f x y z`` where:
+
+	- ``f`` is a *template variable*
+	- ``x``, ``y``, ``z`` are *local binders* (like ``y`` in rule "wombat" above; see definitions).
+	- The arguments ``x``, ``y``, ``z`` are *distinct* variables
+	- ``x``, ``y``, ``z`` must be term variables (not type applications).
+
+* 	A higher order pattern ``f x y z`` matches *any expression* ``e`` provided:
+
+	- The target has the same type as the template
+	- No local binder is free in ``e``, other than ``x``, ``y``, ``z``.
+
+*	If these two condition hold, the higher order pattern ``f x y z`` matches the target expression ``e``, yielding the substitution ``[f :-> \x y z. e]``.
+	Notice that this substitution is type preserving, and the RHS of the substitution has no free local binders.
+
+Uniqueness of matching
+~~~~~~~~~~~~~~~~~~~~~~
+
+Consider this rule and target:
+::
+
+	RULE "funny"   foo (\x y. Just (f x y))
+
+	Target:  ...(foo (\ p q. Just (h (p+1) q)))....
+
+Then during matching we will encounter:
+::
+
+	Template:    f x y
+	Target:      h (p+1) q      [p:->x, q:->y]
+
+The renaming ``[p:->x, q:->y]`` is done by the matcher (today) on the fly, to make the bound variables of the template and target "line up".
+
+Now, we can:
+
+* Either use the new template-application rule to succeed with ``[f :-> \x y. h (x+1) y]``.
+* Or use the existing decompose-application rule to match ``(f x)`` against ``(h (p+1))`` and ``y`` against ``q``.  This will succeed, with ``[f :-> \x. h (x+1)]``.
+
+Critically, *it doesn't matter which we do*.
+We get the same result either way.
+That's encouraging.
+
+More generally, we think that if a match exists it is unique (moudulo eta-reduction).
```

Comment:
Doesn't eta-reduction have perf repercussions? E.g., if `h` has arity 1 there's a difference between the substitutions. Perhaps immaterial to the proposal, but we are bound to think about it in the implementation.